File: api/testing/config_test.go
package testing

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestParseConfig_BasicConfig(t *testing.T) {
	yaml := `
version: 1

settings:
  timeout: 10m
  parallelizable: false

tests:
  - name: basic-test
    description: A basic test
    inputs:
      project.Name: "test"
    steps:
      - block: check-git
        expect: success
    assertions:
      - type: file_exists
        path: generated/README.md
`

	config, err := ParseConfig([]byte(yaml))
	require.NoError(t, err)

	assert.Equal(t, 1, config.Version)
	assert.Equal(t, "10m", config.Settings.Timeout)
	assert.False(t, *config.Settings.Parallelizable)
	require.Len(t, config.Tests, 1)

	tc := config.Tests[0]
	assert.Equal(t, "basic-test", tc.Name)
	assert.Equal(t, "A basic test", tc.Description)
	assert.True(t, tc.Inputs["project.Name"].IsLiteral())
	assert.Equal(t, "test", tc.Inputs["project.Name"].Literal)

	require.Len(t, tc.Steps, 1)
	assert.Equal(t, "check-git", tc.Steps[0].Block)
	assert.Equal(t, StatusSuccess, tc.Steps[0].Expect)

	require.Len(t, tc.Assertions, 1)
	assert.Equal(t, AssertionFileExists, tc.Assertions[0].Type)
	assert.Equal(t, "generated/README.md", tc.Assertions[0].Path)
}

func TestParseConfig_Defaults(t *testing.T) {
	yaml := `
version: 1
tests:
  - name: minimal-test
    steps:
      - block: my-block
`

	config, err := ParseConfig([]byte(yaml))
	require.NoError(t, err)

	// Check defaults
	assert.Equal(t, "5m", config.Settings.Timeout)
	assert.True(t, config.Settings.IsParallelizable())
	assert.Equal(t, StatusSuccess, config.Tests[0].Steps[0].Expect)
}

func TestParseConfig_AllExpectedStatuses(t *testing.T) {
	yaml := `
version: 1
tests:
  - name: status-test
    steps:
      - block: block1
        expect: success
      - block: block2
        expect: fail
      - block: block3
        expect: warn
      - block: block4
        expect: blocked
      - block: block5
        expect: skip
`

	config, err := ParseConfig([]byte(yaml))
	require.NoError(t, err)

	assert.Equal(t, StatusSuccess, config.Tests[0].Steps[0].Expect)
	assert.Equal(t, StatusFail, config.Tests[0].Steps[1].Expect)
	assert.Equal(t, StatusWarn, config.Tests[0].Steps[2].Expect)
	assert.Equal(t, StatusBlocked, config.Tests[0].Steps[3].Expect)
	assert.Equal(t, StatusSkip, config.Tests[0].Steps[4].Expect)
}

func TestParseConfig_AllAssertionTypes(t *testing.T) {
	yaml := `
version: 1
tests:
  - name: assertion-test
    assertions:
      - type: file_exists
        path: file.txt
      - type: file_not_exists
        path: missing.txt
      - type: file_contains
        path: file.txt
        contains: "hello"
      - type: file_not_contains
        path: file.txt
        contains: "goodbye"
      - type: file_matches
        path: file.txt
        pattern: "^hello.*"
      - type: file_equals
        path: file.txt
        value: "exact content"
      - type: output_equals
        block: my-block
        output: result
        value: "expected"
      - type: output_matches
        block: my-block
        output: result
        pattern: "^expected.*"
      - type: output_exists
        block: my-block
        output: result
      - type: files_generated
        block: my-template
        min_count: 1
      - type: script
        command: "test -f output.txt"
      - type: dir_exists
        path: generated/
      - type: dir_not_exists
        path: temp/
`

	config, err := ParseConfig([]byte(yaml))
	require.NoError(t, err)
	require.Len(t, config.Tests[0].Assertions, 13)

	a := config.Tests[0].Assertions
	assert.Equal(t, AssertionFileExists, a[0].Type)
	assert.Equal(t, AssertionFileNotExists, a[1].Type)
	assert.Equal(t, AssertionFileContains, a[2].Type)
	assert.Equal(t, AssertionFileNotContains, a[3].Type)
	assert.Equal(t, AssertionFileMatches, a[4].Type)
	assert.Equal(t, AssertionFileEquals, a[5].Type)
	assert.Equal(t, AssertionOutputEquals, a[6].Type)
	assert.Equal(t, AssertionOutputMatches, a[7].Type)
	assert.Equal(t, AssertionOutputExists, a[8].Type)
	assert.Equal(t, AssertionFilesGenerated, a[9].Type)
	assert.Equal(t, AssertionScript, a[10].Type)
	assert.Equal(t, AssertionDirExists, a[11].Type)
	assert.Equal(t, AssertionDirNotExists, a[12].Type)
}

func TestParseConfig_PerStepAssertions(t *testing.T) {
	yaml := `
version: 1
tests:
  - name: step-assertions
    steps:
      - block: create-resource
        expect: success
        assertions:
          - type: output_exists
            block: create-resource
            output: resource_id
    assertions:
      - type: file_exists
        path: result.txt
`

	config, err := ParseConfig([]byte(yaml))
	require.NoError(t, err)

	require.Len(t, config.Tests[0].Steps[0].Assertions, 1)
	assert.Equal(t, AssertionOutputExists, config.Tests[0].Steps[0].Assertions[0].Type)
	require.Len(t, config.Tests[0].Assertions, 1)
}

func TestParseConfig_FuzzInputs(t *testing.T) {
	yaml := `
version: 1
tests:
  - name: fuzz-test
    inputs:
      project.Name:
        fuzz:
          type: string
          minLength: 10
          maxLength: 20
      project.Count:
        fuzz:
          type: int
          min: 1
          max: 100
`

	config, err := ParseConfig([]byte(yaml))
	require.NoError(t, err)

	// Fuzz inputs are stored as InputValue with Fuzz field
	nameInput := config.Tests[0].Inputs["project.Name"]
	require.NotNil(t, nameInput.Fuzz, "expected project.Name to have fuzz config")
	assert.Equal(t, FuzzString, nameInput.Fuzz.Type)
	assert.Equal(t, 10, nameInput.Fuzz.MinLength)
	assert.Equal(t, 20, nameInput.Fuzz.MaxLength)
}

func TestParseConfig_Cleanup(t *testing.T) {
	yaml := `
version: 1
tests:
  - name: cleanup-test
    cleanup:
      - command: rm -rf /tmp/test
      - path: cleanup/teardown.sh
`

	config, err := ParseConfig([]byte(yaml))
	require.NoError(t, err)

	require.Len(t, config.Tests[0].Cleanup, 2)
	assert.Equal(t, "rm -rf /tmp/test", config.Tests[0].Cleanup[0].Command)
	assert.Equal(t, "cleanup/teardown.sh", config.Tests[0].Cleanup[1].Path)
}

func TestParseConfig_MissingOutputsForBlocked(t *testing.T) {
	yaml := `
version: 1
tests:
  - name: blocked-test
    steps:
      - block: create-resources
        expect: blocked
        missing_outputs:
          - _blocks.create_account.outputs.account_id
`

	config, err := ParseConfig([]byte(yaml))
	require.NoError(t, err)

	step := config.Tests[0].Steps[0]
	assert.Equal(t, StatusBlocked, step.Expect)
	require.Len(t, step.MissingOutputs, 1)
	assert.Equal(t, "_blocks.create_account.outputs.account_id", step.MissingOutputs[0])
}

func TestParseConfig_ValidationErrors(t *testing.T) {
	tests := []struct {
		name        string
		yaml        string
		expectedErr string
	}{
		{
			name: "unsupported version",
			yaml: `
version: 2
tests:
  - name: test
`,
			expectedErr: "unsupported config version: 2",
		},
		{
			name: "no tests",
			yaml: `
version: 1
tests: []
`,
			expectedErr: "at least one test case is required",
		},
		{
			name: "missing test name",
			yaml: `
version: 1
tests:
  - steps:
      - block: test
`,
			expectedErr: "test case 1: name is required",
		},
		{
			name: "missing block in step",
			yaml: `
version: 1
tests:
  - name: test
    steps:
      - expect: success
`,
			expectedErr: `test "test" step 1: block is required`,
		},
		{
			name: "invalid expect status",
			yaml: `
version: 1
tests:
  - name: test
    steps:
      - block: foo
        expect: invalid
`,
			expectedErr: `test "test" step 1: invalid expect value "invalid"`,
		},
		{
			name: "file_exists without path",
			yaml: `
version: 1
tests:
  - name: test
    assertions:
      - type: file_exists
`,
			expectedErr: `test "test" assertion 1: path is required`,
		},
		{
			name: "file_contains without contains",
			yaml: `
version: 1
tests:
  - name: test
    assertions:
      - type: file_contains
        path: file.txt
`,
			expectedErr: `test "test" assertion 1: contains is required`,
		},
		{
			name: "output_equals without block",
			yaml: `
version: 1
tests:
  - name: test
    assertions:
      - type: output_equals
        output: result
        value: test
`,
			expectedErr: `test "test" assertion 1: block is required`,
		},
		{
			name: "invalid timeout",
			yaml: `
version: 1
settings:
  timeout: invalid
tests:
  - name: test
`,
			expectedErr: `invalid timeout format "invalid"`,
		},
		{
			name: "unknown assertion type",
			yaml: `
version: 1
tests:
  - name: test
    assertions:
      - type: unknown_type
        path: file.txt
`,
			expectedErr: `test "test" assertion 1: unknown assertion type "unknown_type"`,
		},
		{
			name: "missing assertion type",
			yaml: `
version: 1
tests:
  - name: test
    assertions:
      - path: file.txt
`,
			expectedErr: `test "test" assertion 1: type is required`,
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			_, err := ParseConfig([]byte(tc.yaml))
			require.Error(t, err)
			assert.Contains(t, err.Error(), tc.expectedErr)
		})
	}
}

func TestGetTimeout(t *testing.T) {
	tests := []struct {
		timeout  string
		expected string
	}{
		{"5m", "5m0s"},
		{"10s", "10s"},
		{"1h", "1h0m0s"},
		{"", "5m0s"}, // default
		{"invalid", "5m0s"}, // falls back to default
	}

	for _, tc := range tests {
		t.Run(tc.timeout, func(t *testing.T) {
			s := TestSettings{Timeout: tc.timeout}
			assert.Equal(t, tc.expected, s.GetTimeout().String())
		})
	}
}

func TestIsParallelizable(t *testing.T) {
	t.Run("default is true", func(t *testing.T) {
		s := TestSettings{}
		assert.True(t, s.IsParallelizable())
	})

	t.Run("explicit true", func(t *testing.T) {
		val := true
		s := TestSettings{Parallelizable: &val}
		assert.True(t, s.IsParallelizable())
	})

	t.Run("explicit false", func(t *testing.T) {
		val := false
		s := TestSettings{Parallelizable: &val}
		assert.False(t, s.IsParallelizable())
	})
}

File: cli/test/config.test.ts
import { describe, it, expect } from "bun:test"
import {
  parseConfig,
  getTimeout,
  getOutputPath,
  isParallelizable,
  shouldUseTempWorkingDir,
  isLiteralInput,
  isFuzzInput,
} from "./config.ts"

const baseYaml = (extra = "") => `version: 1
${extra}
tests:
  - name: smoke
    steps:
      - block: hello-world
        expect: success
`

describe("parseConfig — happy paths", () => {
  it("parses a minimal valid config", () => {
    const cfg = parseConfig(baseYaml())
    expect(cfg.version).toBe(1)
    expect(cfg.tests).toHaveLength(1)
    expect(cfg.tests[0].name).toBe("smoke")
    expect(cfg.tests[0].steps?.[0].block).toBe("hello-world")
    expect(cfg.tests[0].steps?.[0].expect).toBe("success")
  })

  it("defaults the timeout to 5m when missing", () => {
    const cfg = parseConfig(baseYaml())
    expect(cfg.settings.timeout).toBe("5m")
    expect(getTimeout(cfg.settings)).toBe(5 * 60 * 1000)
  })

  it("classifies fuzz vs literal inputs", () => {
    const cfg = parseConfig(`version: 1
tests:
  - name: t
    inputs:
      foo:
        literal: 42
      bar:
        fuzz: { type: string, minLength: 3 }
    steps:
      - block: x
        expect: success
`)
    const inputs = cfg.tests[0].inputs!
    expect(isLiteralInput(inputs.foo)).toBe(true)
    expect(isFuzzInput(inputs.bar)).toBe(true)
    if (isFuzzInput(inputs.bar)) {
      expect(inputs.bar.fuzz.type).toBe("string")
      expect(inputs.bar.fuzz.minLength).toBe(3)
    }
  })

  it("preserves env, description, and step.expect default", () => {
    const cfg = parseConfig(`version: 1
tests:
  - name: t
    description: tag-along desc
    env:
      AWS_REGION: us-east-1
    steps:
      - block: x
`)
    expect(cfg.tests[0].description).toBe("tag-along desc")
    expect(cfg.tests[0].env).toEqual({ AWS_REGION: "us-east-1" })
    // Missing expect defaults to "success".
    expect(cfg.tests[0].steps?.[0].expect).toBe("success")
  })

  it("parses each allowed expect value", () => {
    const cfg = parseConfig(`version: 1
tests:
  - name: t
    steps:
      - { block: a, expect: success }
      - { block: b, expect: fail }
      - { block: c, expect: warn }
      - { block: d, expect: blocked }
      - { block: e, expect: skip }
      - { block: f, expect: config_error }
`)
    const expects = cfg.tests[0].steps!.map((s) => s.expect)
    expect(expects).toEqual([
      "success",
      "fail",
      "warn",
      "blocked",
      "skip",
      "config_error",
    ])
  })

  it("parses assertions with the required fields", () => {
    const cfg = parseConfig(`version: 1
tests:
  - name: t
    steps:
      - block: x
    assertions:
      - { type: file_exists, path: out.txt }
      - { type: file_contains, path: out.txt, contains: hello }
      - { type: output_equals, block: x, output: name, value: alice }
`)
    expect(cfg.tests[0].assertions).toHaveLength(3)
  })
})

describe("parseConfig — failure paths", () => {
  it("rejects an unsupported version", () => {
    expect(() => parseConfig(`version: 2\ntests:\n  - name: x\n    steps: [{block: y}]`)).toThrow(
      /Unsupported config version/,
    )
  })

  it("rejects an empty tests list", () => {
    expect(() => parseConfig(`version: 1\ntests: []`)).toThrow(
      /At least one test case/,
    )
  })

  it("rejects a test case without a name", () => {
    expect(() =>
      parseConfig(`version: 1\ntests:\n  - description: nameless\n    steps: [{block: x}]`),
    ).toThrow(/name is required/)
  })

  it("rejects a step without a block", () => {
    expect(() =>
      parseConfig(`version: 1\ntests:\n  - name: t\n    steps:\n      - expect: success`),
    ).toThrow(/block is required/)
  })

  it("rejects a step with an unknown expect value", () => {
    expect(() =>
      parseConfig(`version: 1\ntests:\n  - name: t\n    steps:\n      - block: x\n        expect: maybe`),
    ).toThrow(/invalid expect value/)
  })

  it("rejects assertions missing required fields per type", () => {
    expect(() =>
      parseConfig(`version: 1\ntests:\n  - name: t\n    steps: [{block: x}]\n    assertions:\n      - type: file_contains\n        path: out.txt`),
    ).toThrow(/contains is required/)

    expect(() =>
      parseConfig(`version: 1\ntests:\n  - name: t\n    steps: [{block: x}]\n    assertions:\n      - type: file_matches\n        path: out.txt`),
    ).toThrow(/pattern is required/)

    expect(() =>
      parseConfig(`version: 1\ntests:\n  - name: t\n    steps: [{block: x}]\n    assertions:\n      - type: output_equals\n        block: x`),
    ).toThrow(/output is required/)
  })

  it("rejects an unknown assertion type", () => {
    expect(() =>
      parseConfig(`version: 1\ntests:\n  - name: t\n    steps: [{block: x}]\n    assertions:\n      - type: misspelled_kind`),
    ).toThrow(/unknown assertion type/)
  })

  it("rejects an unparseable timeout", () => {
    expect(() =>
      parseConfig(`version: 1\nsettings: { timeout: forever }\ntests:\n  - name: t\n    steps: [{block: x}]`),
    ).toThrow(/Invalid timeout format/)
  })
})

describe("settings helpers", () => {
  it("getTimeout parses ms / s / m / h units", () => {
    expect(getTimeout({ timeout: "500ms" })).toBe(500)
    expect(getTimeout({ timeout: "30s" })).toBe(30_000)
    expect(getTimeout({ timeout: "2m" })).toBe(120_000)
    expect(getTimeout({ timeout: "1h" })).toBe(3_600_000)
  })

  it("getOutputPath defaults to 'generated'", () => {
    expect(getOutputPath({})).toBe("generated")
    expect(getOutputPath({ output_path: "out" })).toBe("out")
  })

  it("isParallelizable defaults to true", () => {
    expect(isParallelizable({})).toBe(true)
    expect(isParallelizable({ parallelizable: false })).toBe(false)
  })

  it("shouldUseTempWorkingDir defaults to true", () => {
    expect(shouldUseTempWorkingDir({})).toBe(true)
    expect(shouldUseTempWorkingDir({ use_temp_working_dir: false })).toBe(false)
  })
})
